README.md

```diff
@@ -1,41 +1,124 @@
+```c++
+// Pass additional startup parameters                                                       
+virtual void Start(const std::map<std::string, std::string>& params, 
+                   ray::SpillObjectsCallback spill_objects_callback,
+                   std::function<void()> object_store_full_callback, 
+                   ray::AddObjectCallback add_object_callback, 
+                   ray::DeleteObjectCallback delete_object_callback) = 0;
+```
+
+New virtual functions of  `ObjectStoreRunnerInterface`:
+```c++
+/// The Current total allocated available memory size.
+virtual int64_t GetTotalMemorySize() const = 0;
+
+/// Maximal available memory size. It can be different from the Total memory size if the memory is dynamically expandable.
+virtual int64_t GetMaxMemorySize() const = 0;
+```
+
+The diagram for current plasma object store classes:
+![image](https://github.com/yiweizh-memverge/ray_plugin_manager_rep/blob/main/imgs/original_structure.png)
+
+The diagram for proposed object store plugin classes:
+![image](https://github.com/yiweizh-memverge/ray_plugin_manager_rep/blob/main/imgs/proposed_structure.png)
+
+If the plugin is implemented as a shared library, the implementation should implement an API to create the object store client instance and the runner instance.
+```c++
+extern "C" { 
+    ObjectStoreRunnerInterface CreateRunner(void); 
+    ObjectStoreClientInterface CreateClient(void); 
+}
+```
+### Plugin Manager
+
+The plugin manager shall create the instances that implement `ObjectStoreClientInterface` and `ObjectStoreRunnerInterface` based on specified plugin name if the plugin is implemented within the Ray repository.  If a plugin object store is implemented as a POSIX dynamic linking library, the library shall be opened with its full path using `dlopen()`, `dlsym()` shall be used to look up symbols  for `CreateRunner()` and `CreateClient()` and to lazy create the instances when actually needed. 
+
+The plugin manager shall provide the following APIs:
+```c++
+// Get an instance of the singleton PluginManager
+Static PluginManager& GetInstance()
+
+// Create the object store runner interface with the plugin object store name and configurations
+std::unique_ptr<ObjectStoreRunnerInterface>
+     CreateObjectorStoreRunnerInstance(const std::string& plugin_name,
+   const std::string& plugin_path, 
+   const std::string& plugin_config)
+
+// Create the object store client interface with the plugin object store name and configurations
+std::shared_ptr<ObjectStoreClientInterface> 
+     CreateObjectorStoreClientInstance(const std::string& plugin_name,
+                                       const std::string& plugin_path, 
+                                       const std::string& plugin_config) 
+```
+
+The plugin object store can be specified during Ray startup. The following CLI parameters, or  `ray.init()` options are proposed for the plugin.
+```c++
+ plugin_name:       Optional[str] = ‘default’           
```

Comment:
Probably best to make this part of the [Ray system config](https://github.com/ray-project/ray/blob/master/src/ray/common/ray_config_def.h), which can be overridden with `RAY_<param>` env variables.

Reply:
That's a good idea.